### PR TITLE
DPL: fix rate limiting with QC

### PR DESCRIPTION
**Pull request by ktf**

### CODE REVIEW

**Comment by ktf** [comment]
@davidrohr I assume [a1912aa](https://github.com/AliceO2Group/AliceO2/pull/8563/commits/a1912aa4fe39c75764f76a7802c2a854845b3fd7) is all we need there, no? Not actually tested.

**Comment by davidrohr** [comment]
Ok, it seems I have a lot to test tonight :)

Kind Regards
David Rohr

Sent from my mobile. (Excuse the typos!)

On 12 April 2022 13:32:23 CEST, Giulio Eulisse ***@***.***> wrote:
***@***.*** I assume this is all we need there, no? Not actually tested.

File: Framework/Core/include/Framework/TimesliceIndex.h [hunk 1]
```diff
@@ -115,6 +157,14 @@ class TimesliceIndex
   /// since last time we called getReadyToProcess()
   std::vector<bool> mDirty;
 
+  /// This is the oldest possible timeslice for any given channel
+  /// The cardinality of this vector is the number of input channels
+  std::vector<TimesliceId> mOldestPossibleTimeslices;
+  /// This is the oldest possible timeslice for this index.
+  /// By default we use -1, which means that we don't have any.
+  OldestInputInfo mOldestPossibleInput = {};
```
Comment (knopers8) [inline Framework/Core/include/Framework/TimesliceIndex.h]: unless i am missing something, `0` is the default

**Review by knopers8 — COMMENTED** [review]
To be fair, I do not quite get the general idea of the PR, I would probably need more time to understand it.
I will try to run some QC with it. Now I am rebuilding the whole software stack from scratch, because alibuild took all the remaining 115GBs on my disk :)

File: Framework/Core/include/Framework/TimesliceIndex.h (hunk 1 above)
Comment (ktf) [reply Framework/Core/include/Framework/TimesliceIndex.h]: Yes, this was indeed changed. Never  write documentation.

**Comment by ktf** [comment]
`aliBuild clean` is your friend, assuming you are working with a single branch.

**Comment by knopers8** [comment]
In principle yes, `aliBuild clean` is my friend, but when alibuild runs out of disk space during the build, it usually fails and enters a weird state which I cannot get out in other way than starting the build completely from scratch.

**Comment by ktf** [comment]
I will send you some document with the rationale of the change.

**Comment by ktf** [comment]
The behaviour you report should not actually be the case as everything is meant to be recoverable. If it happens again I would be interested in seeing the actual error of aliBuild clean, since that is probably a bug.

**Comment by knopers8** [comment]
I didn't even try this time, being unsuccessful previous times in such cases. I will make sure to send you an error next time it happens to me.

**Review by knopers8 — DISMISSED** [review]
QC seems to be running as it was when we do not access CCDB. I don't have a reproducer for the issues with QC + CCDB, so I haven't tested that.

**Comment by davidrohr** [comment]
@ktf : Couild you rebase this?

**Comment by ktf** [comment]
Done.

**Comment by ktf** [comment]
@davidrohr how about we limit the rate in the full system test to 3 in fly timeframes, as I guess we would do in production?

**Comment by davidrohr** [comment]
> @davidrohr how about we limit the rate in the full system test to 3 in fly timeframes, as I guess we would do in production?

I think you are somewhat underestimating the EPN :). We have 8 GPUs, and we have overlapped processing on them. Now, since the TFs come one after another, no need to overlap 2 on all GPUs, but I'd say we need 10 TFs to be processed on the GPU. + at least another 10 for the CPU processing that follows + some margin. Then, we have 2 NUMA domains. Perhaps 12 per NUMA domain or so will work. But we have to tune this.

In principle, this should just be a parameter to the readout-proxy, to be set by dpl-workflow.sh. And I believe it is already like this, only it is not set currently, since it led to the workflow getting stuck.

**Comment by ktf** [comment]
I am sure I am underestimating the EPNs, but I meant something different. ;-) Can we do it in the CI? I mean can we enable rate limiting (e.g. in this PR) for the fullCI so that we can test it from now on?

**Comment by davidrohr** [comment]
Ah, I see. Well, perhaps, let's make sure it works before we enable it in the CI. But in any case, the CI runs only 1 single TF, so it will not kick 8n :).

**Comment by ktf** [comment]
This might actually work on linux. I think on mac it currently hangs due to mismatched channel name. Will investigate more tomorrow. The remaining issue seems to be the fact that the DomainInfo header might be received without data, so it does not actually trigger the action of propagating it. This breaks when rate limiting is enabled. This makes sure that whenever we receive some "oldest possible timeframe" notification from upstream we rebroadcast it downstream.

**Comment by ktf** [comment]
The latest incarnation seems to work on mac, provided one specifies the correct name for the ipc file. I will clean it up later today.

**Comment by davidrohr** [comment]
Didn't check yet if it works correctly, but I am getting a huge flood of log messages like:
```
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Done processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 3 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 3 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 3 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 3 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 3 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Done processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 2 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 2 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 2 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 2 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 2 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Done processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 1 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 1 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 1 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 1 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 1 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Done processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 0 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 0 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 0 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 0 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Marking slot 0 as done
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
[24953:internal-dpl-injected-dummy-sink]: [14:25:47][INFO] Done processing timeslice:18446744073709551615, tfCounter:4294967295, firstTFOrbit:4294967295, action:0
```

This is without running, just having the readout proxy and a small workflow idling.

Command to reproduce:
```
O2_NO_CATCHALL_EXCEPTIONS=1 GLOBALDPLOPT="--no-batch" DPL_REPORT_PROCESSING=1 ENABLE_METRICS=1 GEN_TOPO_WORKDIR=/home/qon/tmp5/ QC_REDIRECT_MERGER_TO_LOCALHOST=1 WORKFLOW_PARAMETERS=NONE WORKFLOW_DETECTORS_QC=FT0,FV0,FDD,ZDC,TRD,TOF,ITS,MFT,MID,MCH,EMC,PHS,CPV WORKFLOW_DETECTORS=ALL SYNCMODE=1 SHMSIZE=$((12<<30)) EXTINPUT=1 WORKFLOW_DETECTORS=FT0 TIMEFRAME_RATE_LIMIT=5 ~/alice/O2/prodtests/full-system-test/dpl-workflow.sh
```

**Comment by ktf** [comment]
I do not understand:

```
→ git grep "Marking slot"
→ echo $?
1
```

**Comment by ktf** [comment]
@davidrohr Latest update as we discussed. I will investigate how to fix the 1-timeframe extra latency, but we should simply go ahead with this PR when you are happy with it.

**Comment by ktf** [comment]
Updated. I should probably revert the o2DiamondWorkflow.cxx, but if this converges before I will be able to do it this evening, please go ahead and merge it.

**Comment by ktf** [comment]
This was all green, @davidrohr also tried on the EPN and it works fine with rate limiting disabled, so I am merging this as it fixes for sure some rate limiting setups. I will continue the debugging from here.

**Pull request merged**